### PR TITLE
Search also the email and displayname in user mangement for groups

### PR DESCRIPTION
**Pull request by MorrisJobke**
Similar to #17800 but also searches the email address.

This streamlines the search in groups with the one in users: 

https://github.com/nextcloud/server/blob/145eee91fe611a9ae59c5e6d9a8d06104112c1a5/lib/private/User/Database.php#L266-L278

See #13951

* [x] when testing this it doesn't show the results - but executing the query by hand finds the correct users. Any idea what causes the trouble here?

### CODE REVIEW

**Comment by blizzz** [comment]
> when testing this it doesn't show the results - but executing the query by hand finds the correct users. Any idea what causes the trouble here?

WFM :tm:

**Review by blizzz — APPROVED** [review]

**Comment by MorrisJobke** [comment]
> [x] when testing this it doesn't show the results - but executing the query by hand finds the correct users. Any idea what causes the trouble here?

Works for me now.

Unfortunately the unified search made the search field unavailable for the user management. Thus I based this on commit 7b91cb5029 which is the one right before the unified search got merged. Then we can still test this and work on a solution for this in 20.

@skjnldsv Is this the wanted behavior of the unified search? Because some apps (like the user management here or the app management) use this for filtering the current view.

@juliushaertl @nickvergessen @kesselb Mind to review this?

@blizzz Does a backport make sense here?

**Comment by skjnldsv** [comment]
> @skjnldsv Is this the wanted behavior of the unified search? Because some apps (like the user management here or the app management) use this for filtering the current view.

We're discussing this.
The apps are going to appear in the unified search https://github.com/nextcloud/server/pull/22102
For the users it's up to @jancborchardt

**Comment by MorrisJobke** [comment]
> For the users it's up to @jancborchardt

Same for app management, right? Because it's an admin task vs an normal use case. But up to @jancborchardt

File: lib/private/Group/Database.php [hunk 1]
```diff
@@ -348,15 +348,27 @@ public function usersInGroup($gid, $search = '', $limit = -1, $offset = 0) {
...
 
 		if ($search !== '') {
-			$query->andWhere($query->expr()->like('uid', $query->createNamedParameter(
-				'%' . $this->dbConn->escapeLikeParameter($search) . '%'
-			)));
+			$query->leftJoin('g', 'users', 'u', $query->expr()->eq('g.uid', 'u.uid'))
+				->leftJoin('u', 'preferences', 'p', $query->expr()->andX(
+					$query->expr()->eq('userid', 'u.uid'),
+					$query->expr()->eq('appid', $query->expr()->literal('settings')),
+					$query->expr()->eq('configkey', $query->expr()->literal('email')))
```
Comment (nickvergessen) [inline lib/private/Group/Database.php:361]:
```suggestion
					$query->expr()->eq('p.userid', 'u.uid'),
					$query->expr()->eq('p.appid', $query->expr()->literal('settings')),
					$query->expr()->eq('p.configkey', $query->expr()->literal('email')))
```

File: lib/private/Group/Database.php [hunk 2]
```diff
@@ -348,15 +348,27 @@ public function usersInGroup($gid, $search = '', $limit = -1, $offset = 0) {
...
+				)
+				// sqlite doesn't like re-using a single named parameter here
+				->andWhere(
+					$query->expr()->orX(
+						$query->expr()->ilike('g.uid', $query->createNamedParameter('%' . $this->dbConn->escapeLikeParameter($search) . '%')),
+						$query->expr()->ilike('displayname', $query->createNamedParameter('%' . $this->dbConn->escapeLikeParameter($search) . '%')),
+						$query->expr()->ilike('configvalue', $query->createNamedParameter('%' . $this->dbConn->escapeLikeParameter($search) . '%'))
```
Comment (nickvergessen) [inline lib/private/Group/Database.php:368]:
```suggestion
						$query->expr()->ilike('u.displayname', $query->createNamedParameter('%' . $this->dbConn->escapeLikeParameter($search) . '%')),
						$query->expr()->ilike('p.configvalue', $query->createNamedParameter('%' . $this->dbConn->escapeLikeParameter($search) . '%'))
```

File: lib/private/Group/Database.php [hunk 3]
```diff
@@ -348,15 +348,27 @@ public function usersInGroup($gid, $search = '', $limit = -1, $offset = 0) {
...
+					)
+				)
+				->orderBy('uid_lower', 'ASC');
```
Comment (nickvergessen) [inline lib/private/Group/Database.php:371]:
```suggestion
				->orderBy('u.uid_lower', 'ASC');
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]
Can we have a dedicated test added?

**Review by juliusknorr — APPROVED** [review]

**Comment by MorrisJobke** [comment]
> Can we have a dedicated test added?

I added some integration tests to verify the outside view of this. Otherwise I need to have mocked so much, that it barely is useful in the future.

File: build/integration/features/provisioning-v2.feature [hunk 4]
```diff
@@ -19,3 +19,13 @@ Feature: provisioning
     Then the OCS status code should be "998"
     And the HTTP status code should be "404"
 
+  Scenario: Searching by displayname in groups
+    Given As an "admin"
+    And user "user-in-group" with displayname "specific-name" exists
+    And user "user-in-group2" with displayname "another-name" exists
+    And group "group-search" exists
+    When sending "GET" to "/cloud/users/detailsoffset=25&limit=25&search="fic-n"
```
Comment (nickvergessen) [inline build/integration/features/provisioning-v2.feature:27]: :confused: 

```suggestion
    When sending "GET" to "/cloud/users/detailsoffset=25&limit=25&search=fic-n"
```

Comment (MorrisJobke) [reply build/integration/features/provisioning-v2.feature:27]: 🙈

**Comment by MorrisJobke** [comment]
> I added some integration tests to verify the outside view of this. Otherwise I need to have mocked so much, that it barely is useful in the future.

I refined them a bit and they work locally. Also did some negative testing locally so that they actually indicate when something breaks and it is the case.

Ready for a final review :)

**Comment by jancborchardt** [comment]
Yep, so there we hit the other part of what we discussed – unified search works great overall, but it does have drawbacks when you are inside an app itself. There e.g. in Talk, Users and Apps, direct filtering of the content is nicer.

This is why in the spec we discussed that the apps which already have in-app search which do proper filtering:
- Can move their in-app search/filter to the top of the navigation, just like Talk – this would work nicely here for User management as well as App management
- Still get the benefit of the unified search in the top right.

Makes sense?

**Review by nickvergessen — APPROVED** [review]

**Comment by MorrisJobke** [comment]
> This is why in the spec we discussed that the apps which already have in-app search which do proper filtering:
> 
> * Can move their in-app search/filter to the top of the navigation, just like Talk – this would work nicely here for User management as well as App management
> * Still get the benefit of the unified search in the top right.
> 
> Makes sense?

How to differentiate then between the filter and the unified search? Or should we maybe move the filtering from the header into the actual app content?

**Comment by jancborchardt** [comment]
> How to differentiate then between the filter and the unified search? Or should we maybe move the filtering from the header into the actual app content?

Yep, as said:

> Can move their in-app search/filter to the top of the navigation, just like Talk

;)

**Comment by MorrisJobke** [comment]
> > Can move their in-app search/filter to the top of the navigation, just like Talk

Ah - I thought adding it to the "top navigation" (aka header) 🙈

**Comment by blizzz** [comment]
> @blizzz Does a backport make sense here?

yes (:

**Comment by blizzz** [comment]
> in Talk, Users and Apps, direct filtering of the content is nicer.

at least in Users search has to go via the server. Just too much to pull all of them into the browser.

**Pull request merged**